### PR TITLE
Update HASH Sidebar

### PR DESCRIPTION
**Pull request by teenoh**
## 🌟 What is the purpose of this PR?

This PR includes the initial implementation of the new design for the `PageSidebar` component and updates the existing MUI setup where necessary

## ⚠️ Known issues

- Page menu UI isn't implemented yet
- Type menu UI isn't completely implemented yet
- Bookmarks section
- Drag/Drop/Reordering functionality in pages section not implemented
- Shortcut commands not added to tooltips
- Workspace switcher doesn't get updated with the right details when user switches between workspaces
- Added comments with prefix `@todo-mui` for issues I plan to revisit

## 🐾 Next steps

- [ ] Update Workspace switcher with the right details when user switches between workspaces
- [ ]  Complete EntityType menu UI
- [ ]  Complete Page menu UI
- [ ] Review and fix all broken hover/focus states
- [ ] Add styles for button variants

## 🔍 What does this change?

- Adds an icon for toggling sidebar and a SidebarContextProvider for tracking sidebar open state
- Adds an Avatar component
- Renames the existing `Button` component to `OldButton` and adds a Button component which should be our default. The `OldButton` component will be removed once we've replaced it's references with the new Button
- Renamed CreatePage -> CreatePageModal
- Removed existing css module styles in CreatePage and implemented the styles with MUI
-  Updated `useAccountPages` to call `getAccountPagesTree` mutation. This is useful in providing the necessary info to render pages in a tree view in the sidebar
- Removed css modules styles in `MainContentWrapper` and implemented styles with MUI
- Added a `TopNavLink` component for the navlinks on the sidebar that aren't expandable
- Added a `NavLink` component for expandable navlinks on the sidebar (i.e Pages, Types sections)
- Added a `WorkspaceSwitcher` component for switching between user and user's orgs workspaces
- Added a Types section that renders entity types in a list and has a button for creating an entity type
- Added a Pages section that renders the pages in a tree view and has a button for creating a page.
- Updated the login button in the header to navigate to the login page instead of bringing up a modal because of a possible z-index conflict with the sidebar

### 📜 Does this require a change to the docs?

- No

## 🔗 Related links

- [Asana task](link) (_internal_)

## 🛡 What tests cover this?

- Manual

## ❓ How to test this?

1.  Checkout the branch
2.  Login and compare sidebar design to what's in the ui

## 📹 Demo

https://user-images.githubusercontent.com/16928688/156763146-ed368984-4c0e-4cac-a1b8-3a4e6800c3b8.mp4

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201472293975847

### CODE REVIEW

File: packages/hash/frontend/src/graphql/queries/account.queries.ts
Comment (thehabbos007) [inline packages/hash/frontend/src/graphql/queries/account.queries.ts:51]: This is now `parentPageEntityId` on main

File: packages/hash/frontend/src/graphql/queries/account.queries.ts
Comment (teenoh) [reply packages/hash/frontend/src/graphql/queries/account.queries.ts]: Thanks!

File: packages/hash/frontend/src/graphql/queries/account.queries.ts
Comment (teenoh) [reply packages/hash/frontend/src/graphql/queries/account.queries.ts]: Thanks!

File: packages/hash/frontend/src/components/Button.tsx
Comment (teenoh) [inline packages/hash/frontend/src/components/Button.tsx:8]: Pulled this over from BP repo. Actual styles/variants will be implemented in this [internal link](https://app.asana.com/0/1201864704093982/1201916223398333)

File: packages/hash/frontend/src/components/Link.tsx
Comment (teenoh) [inline packages/hash/frontend/src/components/Link.tsx:94]: Removed this @akash-joshi because this was an override for the hover styles applied by the previous global styles config. We've removed that and as a result no longer need this

File: packages/hash/frontend/src/components/Link.tsx
Comment (teenoh) [inline packages/hash/frontend/src/components/Link.tsx:118]: see https://github.com/hashintel/hash/pull/367/files#r819516130

File: packages/hash/frontend/src/theme/components/dataDisplay/MuiIconButtonThemeOptions.ts
Comment (teenoh) [inline packages/hash/frontend/src/theme/components/dataDisplay/MuiIconButtonThemeOptions.ts:15]: Had to update this so it matches more with the design. This will be revisited properly (including the focus styles) when [internal link](https://app.asana.com/0/1201864704093982/1201916223398333) is worked on

File: packages/hash/frontend/src/graphql/queries/user.queries.ts
Comment (teenoh) [inline packages/hash/frontend/src/graphql/queries/user.queries.ts:21]: use this to determine the number of people in each org

File: packages/hash/frontend/src/components/forms/OldButton.tsx
Comment (benwerner01) [inline packages/hash/frontend/src/components/forms/OldButton.tsx:3]: I'm not sure if this makes sense in react, but could add a `@deprecated` field for this

File: packages/hash/frontend/src/components/layout/PageSidebar/AccountEntityTypeList/EntityTypeMenu.tsx
Comment (benwerner01) [inline packages/hash/frontend/src/components/layout/PageSidebar/AccountEntityTypeList/EntityTypeMenu.tsx:49]: I think this should be open on first render
```suggestion
  const [open, setOpen] = useState(true);
```

**Review by benwerner01 — APPROVED** [review]
Thanks Teeno, I've tested this out locally and it works as intended (taking into consideration the known issues listed in the PR description).

As this PR sets out to be an initial implementation of the sidebar, doesn't remove any existing functionality from the editor (as far as I can tell), and also implements changes to the base theme that other branches are depending on, I'm happy for this to be merged as is.

I'm going to leave my more in-depth review of the code until after this PR has been merged, so that this doesn't block merging when they're going to be addressed in a follow-up PR anyway.

File: packages/hash/frontend/src/components/layout/MainContentWrapper.tsx
Comment (benwerner01) [inline packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:69]: I think this should now have a maximum height, so that you can't scroll past the sidebar:
image

When the content overflows, it should have a scrollbar.

Comment (teenoh) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:69]: Thanks for catching, will update

File: packages/hash/frontend/src/components/layout/MainContentWrapper.tsx
Comment (teenoh) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: Thanks!

Comment (teenoh) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx]: Fixed in ee7fa67

File: packages/hash/frontend/src/components/layout/PageHeader/ActionsDropdown.tsx
Comment (teenoh) [inline packages/hash/frontend/src/components/layout/PageHeader/ActionsDropdown.tsx:73]: @aj had to set this here because I noticed we are using IconButtons with different hover/focus styles. As a result placing default styles in MuiIconButtonThemeOptions caused some issues. This is a temporary fix, but should be properly addressed in this task [internal link](https://app.asana.com/0/1201864704093982/1201916223398333)

File: packages/hash/frontend/src/components/Avatar.tsx
Comment (benwerner01) [inline packages/hash/frontend/src/components/Avatar.tsx:36]:
```
{title.charAt(0).toUpperCase()}
```

File: packages/hash/frontend/src/components/layout/MainContentWrapper.tsx
Comment (benwerner01) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:72]: Awesome, thanks 👍

File: packages/hash/frontend/src/components/layout/MainContentWrapper.tsx
Comment (benwerner01) [inline packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: @teenoh @akash-joshi fun fact, the following (equivalent) syntax exists:
```TypeScript
<Box display="flex" position="relative">
```
Using these props can significantly reduce the number of JSX code lines, and is also more readable because there's no `sx={{`.

File: packages/hash/frontend/src/components/layout/MainContentWrapper.tsx
Comment (benwerner01) [inline packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:46]: @MaggieAppleton correct me if I'm wrong, but my guess is we want these icons to be at the same height:

https://user-images.githubusercontent.com/42802102/156800315-399f74af-00b7-4765-8fb8-a1efa06f6091.mov

File: packages/hash/frontend/src/components/layout/PageSidebar/AccountPageList/CreatePageButton.tsx
Comment (benwerner01) [reply packages/hash/frontend/src/components/layout/PageSidebar/AccountPageList/CreatePageButton.tsx:32]: Also ideally we wouldn't have separate `CreatePageButton` and `CreateEntityTypeButton` components, as their implementation is near identical. If the modal were to be removed from `CreateEntityTypeButton` they would be identical.

File: packages/hash/frontend/src/components/layout/PageSidebar/WorkspaceSwitcher.tsx
Comment (benwerner01) [inline packages/hash/frontend/src/components/layout/PageSidebar/WorkspaceSwitcher.tsx:265]: I think this should be a MUI `List` where the items are `ListItemButton`s. This would match the implementation of the `AccountDropdown` in the header.

Also currently the spacing here is too large (would like be fixed by using `ListItemButton`):
image

File: packages/hash/frontend/src/components/layout/MainContentWrapper.tsx
Comment (akash-joshi) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: @benwerner01 arguably I would say this is more unreadable. CSS is generally grouped together, either in a CSS declaration, a `style` prop, or an `sx` prop in Mui's case.

In my opinion, the reduced number of lines of code aren't worth the extra time required by devs to grok the code (although I appreciate in this case it's pretty straightforward because there aren't any other props on the component).

Comment (benwerner01) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: @akash-joshi didn't we also use `classname="flex"` when we were using tailwind? Would you have had the same objections then?

I think we might have to agree to disagree on this - so happy for this not to be changed in that case, as it does result in equivalent functionality.

However I'd argue that because we  agreed to use MUI, we shouldn't shy away from using MUI-specific convenience features (such as these props for the `Box` component), just because they look slightly different than the default way of doing things. I strongly disagree that this would "grok" the code. It's a feature of a library we agreed to use - so it's not unreasonable to expect developers to be able to read/use it (just as we'd expect them to use the MUI-specific `sx` prop, instead of putting things into the default `style` prop). In this case it's also really straightforward what a `position="relative"` prop would do, so I don't see how this would introduce any readability issues.

And I do think that the size of react trees can quickly become a readability issue in MUI when we're always chucking things like `display: "flex"` into the `sx` prop every time we want to use it, where it has to take up a whole line in the react tree. I'm pretty sure that's what the thinking behind adding these `Box` props was, because it let's these common styles be defined together in one line. 

I guess my thoughts boil down to: we agreed to use MUI, let's try and use it to it's full potential.

Comment (nonparibus) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: @akash-joshi I find myself strongly agreeing with Ben on this. We have consciously bought into MUI now, so let's utilize their style fully.

Comment (teenoh) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: I think the benefit of using the inline syntax disappears once we have more than 3 styles.

Prettier will format this

```
 <Box display="flex" position="absolute" justifyContent="center" alignItems="center">
        <div />
 </Box>
```

to this

```
<Box
        display="flex"
        position="absolute"
        justifyContent="center"
        alignItems="center"
      >
        <div />
</Box>
```
which isn't so much different from using the `sx` prop directly. 

Although I understand that placing lots of styles in the `sx` prop can make the component a bit harder reader to read. I think we should adopt using [styled()](https://mui.com/system/styled/) for scenarios where we have a lot of styles in the sx prop

```
const MyComponent = styled('div')(({ theme }) => ({
  display: "flex",
  position: "absolute",
  justifyContent: "center",
  ...
}))


<MyComponent>
   <div />
</MyComponent>
```

We might lose some cool things like the shorthand syntax provided by the sx prop, but we can also use the [experimental_sx prop ](https://mui.com/system/styled/#how-can-i-use-the-sx-syntax-with-the-styled-utility)

Comment (benwerner01) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: @teenoh granted when you add more than 3 of the props they wrap to the next line, but even in that case you're saving yourself two lines of code (`sx={{` and `}}`). And my instinct is we would be using 3 or less of them in most cases (just like in the example which started this thread).

I agree we should make more use of styled components, especially to create more reusable styling. However I don't think that would apply for this example: when all we're trying to do is add `display: "flex"` and `position: "relative"` **we can go from using 6 lines of code to just using 1** - and I think that is something worth doing throughout the codebase.

Comment (benwerner01) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: @teenoh to be clear I agree with the example you gave. I'm thinking of the situations where there isn't a need for the `sx` prop because the inline props can be used instead. When we require styling (such as the `:hover` effect in your example) it doesn't make a difference if `display="flex"` is an inline prop, or defined in `sx`. In the example that started this thread however, it makes a big difference (we would go from 6 lines of code to 1).

File: packages/hash/frontend/src/components/layout/PageSidebar/WorkspaceSwitcher.tsx
Comment (teenoh) [reply packages/hash/frontend/src/components/layout/PageSidebar/WorkspaceSwitcher.tsx:265]: Thanks for catching. I was talking with @akash-joshi earlier and mentioned we should probably be using Menu/MenuItem for UIs like this as opposed to Popover/(Box, ListItemButton) since they are technically menus and Menu/MenuItem handle the accessibility features properly

Comment (teenoh) [reply packages/hash/frontend/src/components/layout/PageSidebar/WorkspaceSwitcher.tsx:265]: Plan to replace existing implementations with Menu/MenuItem in this task => [internal link](https://app.asana.com/0/1201864704093982/1201926880408096)

File: packages/hash/frontend/src/components/layout/PageSidebar/AccountPageList/CreatePageButton.tsx
Comment (teenoh) [reply packages/hash/frontend/src/components/layout/PageSidebar/AccountPageList/CreatePageButton.tsx:32]: Thanks for catching. Noticed this issue too when I was looking over the Header implementation. It also makes use of the modal. Plan to address that in this task => [internal link](https://app.asana.com/0/1201864704093982/1201926880408104). 
Agree we probably wouldn't need to separate the components if this issue is addressed.

File: packages/hash/frontend/src/components/layout/MainContentWrapper.tsx
Comment (teenoh) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: @benwerner01, just to be clear, I agree we should use inline props for components that have less than 4 style rules like in the example above. But for other scenarios, I think we should stick to an approach (inline props or sx/styled) for consistency and I'm in favour of the (sx/styled) approach

Comment (benwerner01) [reply packages/hash/frontend/src/components/layout/MainContentWrapper.tsx:42]: Cool, thanks I think we're on the same page now. I think an easy rule of thumb would be that when we're already using the `sx` for something we shouldn't use inline props (as it makes no difference).

File: packages/hash/frontend/src/components/Avatar.tsx
Comment (benwerner01) [reply packages/hash/frontend/src/components/Avatar.tsx:36]: That wouldn't account for `title = ""` because `??` only checks for `undefined` or `null`

Comment (benwerner01) [reply packages/hash/frontend/src/components/Avatar.tsx:36]: Sorry just saw you updated this. I think in practice it should never be `""` because we require users to have a `preferredName` (which I assume is what we're passing here?). But we can still account for invalid props using `"U"`, so I'm happy with that approach.

**Pull request merged**
